src/MongoDB.Driver.Core/Core/Misc/StreamExtensionMethods.cs

```diff
@@ -36,6 +36,39 @@ public static void EfficientCopyTo(this Stream input, Stream output)
             }
         }
 
+        public static async Task<int> ReadAsync(this Stream stream, byte[] buffer, int offset, int count, TimeSpan timeout, CancellationToken cancellationToken)
```

Comment:
This is the same approach we used to timeout the Connect method which doesn't itself support timeouts.

We **could** do this for `WriteAsync` also but I'm not sure that's needed. Let me know what you think.

Reply:
I think it's needed because, as I think you pointed out to me, a write can suffer from the same issue if the client-side write buffer is filled (say, when doing a large write operation).

Reply:
OK I will do the same for async writes.

Reply:
Done